google-cloud-core/lib/google/cloud/core/environment.rb

```diff
@@ -0,0 +1,141 @@
+          resp = conn.get CHECK_URI do |req|
+            req.options.timeout = 0.1
+          end
+          return false unless resp.status == 200
+          return false unless resp.headers.key? "Metadata-Flavor"
+          return resp.headers["Metadata-Flavor"] == "Google"
+        rescue Faraday::TimeoutError, Faraday::ConnectionFailed
+          return false
+        end
+
+        ##
+        # Check if running from Google Compute Engine, and not using GAE or GKE
+        #
+        # @return [Boolean] True iff running from GCE and not GAE or GKE
+        def self.gce?
+          gce_vm? && !gae? && !gke?
+        end
+
+        ##
+        # Check if running from Google Container Engine by querying for
+        # GKE cluster name and VM instance_zone
+        #
+        # @return [Boolean] True iff self.gke_cluster_name() and
+        # self.instance_zone() both return true values
+        def self.gke?
+          gke_cluster_name && instance_zone
+        end
+
+        ##
+        # Check if running from Google App Engine by checking existance of
+        # GAE_VM, GAE_MODULE_NAME, GAE_MODULE_VERSION environment variables.
+        #
+        # @return [Boolean] True if all three GAE environment variables are
+        #   defined
+        def self.gae?
+          ENV["GAE_VM"] && gae_module_id && gae_module_version
+        end
+
+        def self.project_id
+          uri = "#{PROJECT_URI}/project-id"
+          get_metadata_attribute uri, :project_id
+        end
+
+        ##
+        # Retrieve GAE module name
+        def self.gae_module_id
+          ENV["GAE_MODULE_NAME"]
+        end
+
+        ##
+        # Retrieve GAE module version
+        def self.gae_module_version
+          ENV["GAE_MODULE_VERSION"]
+        end
+
+        ##
+        # Retrieve GKE cluster name
+        def self.gke_cluster_name
+          uri = "#{INSTANCE_METADATA_URI}/attributes/cluster-name"
+          get_metadata_attribute uri, :cluster_name
+        end
+
+        ##
+        # Retrieve GKE namespace id
+        def self.gke_namespace_id
+          ENV["GKE_NAMESPACE_ID"]
+        end
+
+        ##
+        # Retrieve GCE VM zone
+        def self.instance_zone
+          uri = "#{INSTANCE_METADATA_URI}/zone"
+          full_zone = get_metadata_attribute uri, :zone
+          full_zone.nil? ? nil : full_zone.split(%r{/}).last
+        end
+
+        ##
+        # Retrieve GCE VM instance_id
+        def self.instance_id
+          uri = "#{INSTANCE_METADATA_URI}/id"
+          get_metadata_attribute uri, :instance_id
+        end
+
+        ##
+        # Helper method to send HTTP request to GCP metadata service and
+        # retrieve environment information
+        def self.get_metadata_attribute uri, attr_name, connection: nil
+          @metadata ||= {}
+          return @metadata[attr_name] if @metadata.key? attr_name
+
+          conn = connection || Faraday.default_connection
+
+          puts conn.get uri
```

Comment:
Is this `puts` statement left over from debugging during development? Currently, google-cloud has no `puts` output, except for some retries in acceptance tests.